nflow-engine/src/test/java/io/nflow/engine/internal/executor/WorkflowStateProcessorTest.java

```diff
@@ -1078,6 +1102,38 @@ public NextAction start(@SuppressWarnings("unused") StateExecution execution) {
 
   }
 
+  public static class LoopingTestWorkflow extends WorkflowDefinition<LoopingTestWorkflow.State> {
+
+    int counter = 0;
+
+    protected LoopingTestWorkflow() {
+      super("looping-test", State.start, State.error);
+      permit(State.start, State.start);
+      permit(State.start, State.done);
+    }
+
+    public enum State implements WorkflowState {
+      start(WorkflowStateType.start), done(WorkflowStateType.end), error(WorkflowStateType.manual);
+
+      private WorkflowStateType stateType;
+
+      State(WorkflowStateType stateType) {
+        this.stateType = stateType;
+      }
+
+      @Override
+      public WorkflowStateType getType() {
+        return stateType;
+      }
+    }
+
+    public NextAction start(@SuppressWarnings("unused") StateExecution execution) throws InterruptedException {
+      sleep(100);
+      return counter++ < 100 ? moveToState(State.start, "loop") : moveToState(State.done, "done");
```

Comment:
test will fail anyway if looping more than 20 times, wait delay the failure for 80 iterations after that?

Reply:
just to make extra sure some random timing/slowness problem on travis does not break the test. When the test works it takes 1.5 seconds so how long it takes if the code breaks does not matter. At least it is clear when it takes 10 seconds that it was not bad luck.

Reply:
After the counter has reached 20, the test will fail anyway regardless of possible timing issues because of `assertThat(loopingWf.counter, lessThan(20));`
I would actually replace that assert with a check that awaitTermination(5, seconds) returns true (e.g. executor terminated, 5 second timeout should be enough there) and workflow state is `start` (termination was due to shutdown request, not because workflow reached done state).